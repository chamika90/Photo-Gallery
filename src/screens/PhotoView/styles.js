import { StyleSheet } from 'react-native';
import { theme } from 'config/theme';
import { isiOS } from 'helper/utils';

const { colors } = theme;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: colors.primaryImageBackgroundColor,
  },
  imageView: {
    margin: 10,
    backgroundColor: colors.secondaryImageBackgroundColor,
  },
  actionButtonContainer: {
    height: 50,
    width: '100%',
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'flex-end',
  },
  bottomToolBar: {
    position: 'absolute',
    bottom: isiOS ? 30 : 0,
    left: 0,
    height: 60,
    width: '100%',
    flexDirection: 'column',
    alignItems: 'center',
    justifyContent: 'flex-end',
  },
  downloadIconContainer: {
    marginHorizontal: 10,
    padding: 10,
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  downloadIcon: {
    width: 25,
    height: 25,
    tintColor: 'white',
  },
  progressBarContainer: {
    height: 10,
    width: '100%',
    justifyContent: 'flex-end',
  },
});

export default styles;
