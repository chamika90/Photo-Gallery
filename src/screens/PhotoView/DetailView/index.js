import React from 'react';
import {View, Text, TouchableOpacity, Linking} from 'react-native';

import BottomModal from 'components/BottomModal';
import styles from './styles';

const DetailRow = ({title, value}) => {
  return (
    <View style={styles.detailContainer}>
      <Text style={styles.detailTitle}>{title}</Text>
      {(title === 'Web' && (
        <Text
          style={styles.linkText}
          onPress={() => {
            Linking.openURL(value);
          }}>
          {value}
        </Text>
      )) || <Text style={styles.detailValue}>{value}</Text>}
    </View>
  );
};

const DetailView = ({isVisible, item, onClose, onDownloadPress}) => {
  return (
    <BottomModal isVisible={isVisible} onClose={() => onClose()}>
      <View>
        <DetailRow title={'Author'} value={item.author} />
        <DetailRow title={'Web'} value={item.url} />
        <TouchableOpacity
          style={styles.downloadButton}
          onPress={() => onDownloadPress()}>
          <Text style={styles.downloadText}>Download</Text>
        </TouchableOpacity>
      </View>
    </BottomModal>
  );
};

export default DetailView;
