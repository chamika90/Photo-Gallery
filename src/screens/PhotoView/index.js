import React, { useState } from 'react';
import {
  View,
  Dimensions,
  Image,
  TouchableOpacity,
  PermissionsAndroid,
  Platform,
  Alert,
  Share,
} from 'react-native';
import ImageZoom from 'react-native-image-pan-zoom';
import FastImage from 'react-native-fast-image';
import CameraRoll from '@react-native-community/cameraroll';
import RNFetchBlob from 'rn-fetch-blob';
import Toast from 'react-native-toast-message';
import { ProgressBar } from 'react-native-paper';

import styles from './styles';
import Res from 'resources';
import { theme } from 'config/theme';
import DetailView from './DetailView';

const { colors } = theme;
const { Images } = Res;

const ActionButton = ({ icon, onButtonPress }) => {
  return (
    <TouchableOpacity
      style={styles.downloadIconContainer}
      onPress={() => onButtonPress()}>
      <Image style={styles.downloadIcon} source={icon} />
    </TouchableOpacity>
  );
};

/*
 * Photo View Screen
 */
const PhotoView = ({ route }) => {
  const { item } = route.params;

  const aspectRatio = item.height / item.width;
  const screenWidth = Dimensions.get('window').width;
  const screenHeight = Dimensions.get('window').height;

  const imageHeight = aspectRatio * (screenWidth - 20);

  const [progress, setProgress] = useState(0);
  const [loading, setLoading] = useState(false);
  const [isDetailModalVisible, setDetailModalStatus] = useState(false);
  const checkPermissionsAndroid = async () => {
    try {
      const granted = await PermissionsAndroid.request(
        PermissionsAndroid.PERMISSIONS.WRITE_EXTERNAL_STORAGE,
        {
          title: 'Permission',
          message: 'Your permission is required to save images to your device',
          buttonNegative: 'Cancel',
          buttonPositive: 'OK',
        },
      );
      if (granted === PermissionsAndroid.RESULTS.GRANTED) {
        return true;
      }
      Alert.alert(
        'Save remote Image',
        'Please Grant Permission to save Image',
        [{ text: 'OK', onPress: () => console.log('OK Pressed') }],
        { cancelable: false },
      );
    } catch (err) {
      Alert.alert(
        'Save remote Image',
        'Failed to save Image: ' + err.message,
        [{ text: 'OK', onPress: () => console.log('OK Pressed') }],
        { cancelable: false },
      );
    }
  };

  const handleDownload = async () => {
    // check storage permission for android devices
    if (Platform.OS === 'android') {
      const granted = await checkPermissionsAndroid();
      if (!granted) {
        return;
      }
    }

    setLoading(true);
    setProgress(0);

    RNFetchBlob.config({
      fileCache: true,
      appendExt: 'png',
    })
      .fetch('GET', item.download_url)
      .progress((received, total) => {
        setProgress(received / total);
      })
      .then(res => {
        CameraRoll.save(res.data, 'photo')
          .then(res => {
            setLoading(false);
            setProgress(100);
            Toast.show({
              text1: 'Downloaded',
              type: 'success',
            });
          })
          .catch(err => {
            setLoading(false);
            setProgress(0);
            console.log(err);
            Toast.show({
              text1: 'Download Failed',
              type: 'error',
            });
          });
      })
      .catch(error => {
        console.log(error);

        Toast.show({
          text1: 'Download Failed',
          type: 'error',
        });
      });
  };

  const handlePhotoSharing = async () => {
    try {
      const shareOptions = {
        message: item.download_url,
      };
      await Share.share(shareOptions);
    } catch (error) {
      Toast.show({
        text1: 'Failed to share image. Please try again.',
        type: 'error',
      });
    }
  };

  return (
    <>
      <View style={styles.container}>
        <ImageZoom
          cropWidth={screenWidth}
          cropHeight={screenHeight}
          imageWidth={screenWidth}
          imageHeight={imageHeight}>
          <FastImage
            style={[
              styles.imageView,
              {
                width: screenWidth - 20,
                height: imageHeight,
              },
            ]}
            source={{
              uri: item.download_url,
              cache: FastImage.cacheControl.immutable,
              priority: FastImage.priority.normal,
            }}
          />
        </ImageZoom>
        <View style={styles.bottomToolBar}>
          {loading ? (
            <View style={styles.progressBarContainer}>
              <ProgressBar
                color={colors.progressBarColor}
                progress={progress}
              />
            </View>
          ) : null}
          <View style={styles.actionButtonContainer}>
            <ActionButton
              icon={Images.icons.ic_details}
              onButtonPress={() => setDetailModalStatus(true)}
            />

            <ActionButton
              icon={Images.icons.ic_share}
              onButtonPress={() => handlePhotoSharing()}
            />

            <ActionButton
              icon={Images.icons.ic_download}
              onButtonPress={() => handleDownload()}
            />
          </View>
        </View>
      </View>
      <DetailView
        isVisible={isDetailModalVisible}
        item={item}
        onClose={() => {
          setDetailModalStatus(false);
        }}
        onDownloadPress={() => {
          setDetailModalStatus(false);
          handleDownload();
        }}
      />
    </>
  );
};

export default PhotoView;
