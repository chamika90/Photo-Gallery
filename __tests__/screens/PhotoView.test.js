import 'react-native';
import React from 'react';
import {render} from '@testing-library/react-native';

import PhotoView from 'screens/PhotoView';

jest.mock(
  'rn-fetch-blob',
  () => {
    const mRNFetchBlob = {
      fetch: jest.fn(),
    };
    return mRNFetchBlob;
  },
  {virtual: true},
);

describe('PhotoView', () => {
  const mockedParams = {
    route: {
      params: {
        item: {
          author: 'Alejandro Escamilla',
          download_url: 'https://picsum.photos/id/1/5616/3744',
          height: 3744,
          id: '1',
          url: 'https://unsplash.com/photos/LNRyGwIJr5c',
          width: 5616,
        },
      },
    },
    navigation: '',
  };

  it('should match snapshot', async () => {
    const {toJSON} = render(<PhotoView {...mockedParams} />);
    expect(toJSON()).toMatchSnapshot();
  });
});
