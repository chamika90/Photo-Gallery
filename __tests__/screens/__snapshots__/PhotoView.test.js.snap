// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhotoView should match snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#000000",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    onMoveShouldSetResponder={[Function]}
    onMoveShouldSetResponderCapture={[Function]}
    onResponderEnd={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderReject={[Function]}
    onResponderRelease={[Function]}
    onResponderStart={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    onStartShouldSetResponderCapture={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "transparent",
        "height": 1334,
        "justifyContent": "center",
        "overflow": "hidden",
        "width": 750,
      }
    }
  >
    <View
      renderToHardwareTextureAndroid={true}
      style={
        Object {
          "transform": Array [
            Object {
              "scale": 1,
            },
            Object {
              "translateX": 0,
            },
            Object {
              "translateY": 0,
            },
          ],
        }
      }
    >
      <View
        onLayout={[Function]}
        style={
          Object {
            "height": 486.66666666666663,
            "width": 750,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Array [
                Object {
                  "margin": 10,
                },
                Object {
                  "height": 486.66666666666663,
                  "width": 730,
                },
              ],
            ]
          }
        >
          <FastImageView
            resizeMode="cover"
            source={
              Object {
                "cache": "immutable",
                "priority": "normal",
                "uri": "https://picsum.photos/id/1/5616/3744",
              }
            }
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
          />
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "bottom": 30,
        "flexDirection": "column",
        "height": 60,
        "justifyContent": "flex-end",
        "left": 0,
        "position": "absolute",
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "height": 50,
          "justifyContent": "flex-end",
          "width": "100%",
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "marginHorizontal": 10,
            "opacity": 1,
            "padding": 10,
          }
        }
      >
        <Image
          source={
            Object {
              "testUri": "../../../src/resources/images/icons/icShare/ic_share.png",
            }
          }
          style={
            Object {
              "tintColor": "white",
            }
          }
        />
      </View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "marginHorizontal": 10,
            "opacity": 1,
            "padding": 10,
          }
        }
      >
        <Image
          source={
            Object {
              "testUri": "../../../src/resources/images/icons/icDownload/ic_download.png",
            }
          }
          style={
            Object {
              "tintColor": "white",
            }
          }
        />
      </View>
    </View>
  </View>
</View>
`;
